cmd/login_test.go

```diff
@@ -0,0 +1,40 @@
+package cmd
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/applandinc/appland-cli/internal/appland"
+	"github.com/stretchr/testify/mock"
+)
+
+type MockClient struct {
+	mock.Mock
+	appland.Client
+}
+
+func (m *MockClient) Login(login string, password string) error {
+	args := m.Called(login, password)
+	return args.Error(0)
+}
+
+func TestLoginCommand(t *testing.T) {
+
+	testClient := &MockClient{}
+
+	testClient.On("Login", "bob", "secret").Return(nil)
```

Comment:
I'd prefer to see the HTTP request mocked instead of the client because this keeps us from having to structure our code around our tests. Something like:

```go
defer gock.Off()
gock.New("http://localhost:3000").
	Post("/api/api_keys").
	MatchHeader("Authorization", "Basic Ym9iOnNlY3JldA==").
	MatchType("json").
	Reply(200).
	JSON(map[string]string{"api_key": "3977c1ca-6d87-49ae-bac7-97a9440a0149"})
```

See the Client tests for more examples of this.

Reply:
I considered this. It has the effect of bringing the implementation details of the client up into the commands, which I see as significant downside.

Dependency injection is a tried and true mechanism for making things testable. Restructuring the code to use it doesn't seem like a bad thing.

Reply:
That's fair. At the very least let's remove or export the non-exported methods from the interface.